### PR TITLE
Fix Translog.Delete serialization for sequence numbers

### PR DESCRIPTION
**Pull request by dakrone**
Translog.Delete used `.writeVLong` instead of `.writeLong` for the sequence
number and primary term (and their respective "read" variants). This could lead
to issues where a 5.x node sent a translog operation with a negative sequence
number (-2 for unassigned seq no) that tripped an assertion serializing a
negative number and causing ES to exit.

Adds a unit test for serialization and a mixed-cluster REST test, since that was
how this was originally caught.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 1]
```diff
@@ -1993,4 +2003,46 @@ public void testPendingDelete() throws IOException {
     public static Translog.Location randomTranslogLocation() {
         return new Translog.Location(randomLong(), randomLong(), randomInt());
     }
+
+    public void testTranslogOpSerialization() throws Exception {
+        BytesReference B_1 = new BytesArray(new byte[]{1});
+        SeqNoFieldMapper.SequenceID seqID = SeqNoFieldMapper.SequenceID.emptySeqID();
+        long randomSeqNum = randomLong();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Can we not make these `randomLong`, instead make them either `randomNonNegativeLong` or `SequenceNumberService.UNASSIGNED_SEQ_NO`?

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 2]
```diff
@@ -1993,4 +2003,46 @@ public void testPendingDelete() throws IOException {
...
+        long randomPrimaryTerm = randomLong();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Similar thing here, can we use a realistic primary term (they start one and will default to zero for BWC)?

**Review by jasontedor — APPROVED** [review]
Thanks @dakrone. I left two comments about using realistic values but otherwise LGTM.

**Review by jasontedor — APPROVED** [review]
Still LGTM.

**Comment by dakrone** [comment]
retest this please

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 3]
```diff
@@ -1993,4 +2003,46 @@ public void testPendingDelete() throws IOException {
     public static Translog.Location randomTranslogLocation() {
         return new Translog.Location(randomLong(), randomLong(), randomInt());
     }
+
+    public void testTranslogOpSerialization() throws Exception {
+        BytesReference B_1 = new BytesArray(new byte[]{1});
+        SeqNoFieldMapper.SequenceID seqID = SeqNoFieldMapper.SequenceID.emptySeqID();
+        long randomSeqNum = randomBoolean() ? SequenceNumbersService.UNASSIGNED_SEQ_NO : randomNonNegativeLong();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: can we add a comment this is only relevant for a BWC perspective and can be removed in ES 7.0?

**Comment by bleskes** [comment]
LGTM2. good catch

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java (hunk 3 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: I don't understand, wouldn't we want to test translog operation serialization for all versions?

**Comment by dakrone** [comment]
retest this please

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java (hunk 3 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: yes. I should have been more precise - the usage of SequenceNumbersService.UNASSIGNED_SEQ_NO is only relevant due to a BWC aspect.

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: The sequence number backwards layer won't be needed after 7.0.0

Comment (dakrone) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Ahh gotcha, I'll add a comment then.

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Asserts are better for this. 😄

Comment (dakrone) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: I changed the comment to an assert.

**Review by jasontedor — APPROVED** [review]
LGTM.

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java (hunk 3 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: Thanks @dakrone.

**Pull request merged**